core/Enclave/enclave_self_test.cpp

```diff
@@ -0,0 +1,939 @@
+    {
+        if (!aes_gcm_encryption(test_vector))
+        {
+            log_e("fail encrypt at %s case %d\n", __FUNCTION__, index);
+            continue;
+        }
+
+        if (!aes_gcm_decryption(test_vector))
+        {
+            log_e("fail decrypt at %s case %d\n", __FUNCTION__, index);
+            continue;
+        }
+
+        index++;
+    }
+
+    if (index != aes_gcm_crypto_test_vectors.size() + 1)
+    {
+        return false;
+    }
+
+    return true;
+}
+
+bool sm4_crypto_test()
+{
+    int index = 1;
+
+    for (auto test_vector : sm4_ctr_crypto_test_vectors)
+    {
+        if (!sm4_ctr_encryption(test_vector))
+        {
+            log_e("fail encrypt at %s case %d\n", __FUNCTION__, index);
+            continue;
+        }
+
+        if (!sm4_ctr_decryption(test_vector))
+        {
+            log_e("fail decrypt at %s case %d\n", __FUNCTION__, index);
+            continue;
+        }
+
+        index++;
+    }
+    for (auto test_vector : sm4_cbc_crypto_test_vectors)
+    {
+        if (!sm4_cbc_encryption(test_vector))
+        {
+            log_e("fail encrypt at %s case %d\n", __FUNCTION__, index);
+            continue;
+        }
+
+        if (!sm4_cbc_decryption(test_vector))
+        {
+            log_e("fail decrypt at %s case %d\n", __FUNCTION__, index);
+            continue;
+        }
+
+        index++;
+    }
+
+    if (index != sm4_ctr_crypto_test_vectors.size() + sm4_cbc_crypto_test_vectors.size() + 1)
+    {
+        return false;
+    }
+
+    return true;
+}
+
+static EC_GROUP *create_EC_group(const char *p_hex, const char *a_hex,
+                                 const char *b_hex, const char *x_hex,
+                                 const char *y_hex, const char *order_hex,
+                                 const char *cof_hex)
+{
+    BIGNUM *p = NULL;
+    BIGNUM *a = NULL;
+    BIGNUM *b = NULL;
+    BIGNUM *g_x = NULL;
+    BIGNUM *g_y = NULL;
+    BIGNUM *order = NULL;
+    BIGNUM *cof = NULL;
+
+    EC_POINT *generator = NULL;
+    EC_GROUP *group = NULL;
+
+    int ok = 0;
+
+    if (!BN_hex2bn(&p, p_hex) ||
+        !BN_hex2bn(&a, a_hex) ||
+        !BN_hex2bn(&b, b_hex))
+        goto done;
+
+    group = EC_GROUP_new_curve_GFp(p, a, b, NULL);
+    generator = EC_POINT_new(group);
+    if (!BN_hex2bn(&g_x, x_hex) ||
+        !BN_hex2bn(&g_y, y_hex) ||
+        !EC_POINT_set_affine_coordinates(group, generator, g_x, g_y, NULL))
+        goto done;
+
+    if (!BN_hex2bn(&order, order_hex) ||
+        !BN_hex2bn(&cof, cof_hex) ||
+        !EC_GROUP_set_generator(group, generator, order, cof))
+        goto done;
+    ok = 1;
+
+done:
+    BN_free(p);
+    BN_free(a);
+    BN_free(b);
+    BN_free(g_x);
+    BN_free(g_y);
+    EC_POINT_free(generator);
+    BN_free(order);
+    BN_free(cof);
+    if (!ok)
+    {
+        EC_GROUP_free(group);
+        group = NULL;
+    }
+    return group;
+}
+
+static bool sm2_verify_test(map<string, string> test_vector)
```

Comment:
add some comments for each test cases, especially descript the format of the test_vector which is used

Reply:
ok